File: src/shm_bridge_cpp/src/generic_shm_writer.cpp
// generic_shm_writer.cpp — type-agnostic ROS 2 -> shared memory writer (PROMPT.md §5).
//
// Uses rclcpp::GenericSubscription so it can subscribe to ANY message type by
// name at runtime (no compile-time type dependency). Every message is written via
// the CDR path (universal). The Python writer covers the FLAT zero-copy fast path
// for image-like types; this C++ node is the universal-transport counterpart and
// shares the exact same binary header + recipe contract.
//
// Parameters:
//   streams (string[])  "stream_name:topic:Type" e.g.
//        ["tf:/tf:tf2_msgs/msg/TFMessage", "rgb:/rgb:sensor_msgs/msg/Image"]
//   max_bytes (int)     per-stream frame cap (default 1920*1080*4)
#include "shm_bridge_cpp/shm_contract.hpp"

#include <rclcpp/rclcpp.hpp>
#include <rclcpp/generic_subscription.hpp>
#include <rclcpp/serialized_message.hpp>

#include <chrono>
#include <fstream>
#include <memory>
#include <sstream>
#include <string>
#include <vector>

using namespace shm_contract;

struct Stream {
    std::string name, topic, type_name, recipe_path;
    void* hdr = nullptr;
    void* frame = nullptr;
    size_t frame_cap = 0;
    uint32_t seq = 0;
    rclcpp::GenericSubscription::SharedPtr sub;
};

static std::vector<std::string> split3(const std::string& s) {
    // split on first two ':' -> name, topic, type
    auto a = s.find(':');
    auto b = s.find(':', a + 1);
    return {s.substr(0, a), s.substr(a + 1, b - a - 1), s.substr(b + 1)};
}

class GenericSHMWriter : public rclcpp::Node {
public:
    GenericSHMWriter() : Node("generic_shm_writer_cpp") {
        declare_parameter<std::vector<std::string>>(
            "streams", {"tf:/tf:tf2_msgs/msg/TFMessage"});
        declare_parameter<int>("max_bytes", 1920 * 1080 * 4);

        max_bytes_ = static_cast<size_t>(get_parameter("max_bytes").as_int());
        auto specs = get_parameter("streams").as_string_array();

        for (auto& spec : specs) {
            auto parts = split3(spec);
            auto st = std::make_shared<Stream>();
            st->name = parts[0];
            st->topic = parts[1];
            st->type_name = parts[2];
            st->recipe_path = recipe_path(st->name);

            size_t hs = 0, fs = 0;
            st->hdr = map_file(header_path(st->name), HEADER_SIZE, true, hs);
            st->frame = map_file(frame_path(st->name), max_bytes_, true, fs);
            st->frame_cap = fs;

            Header h{};
            set_type_name(h, st->type_name);
            std::memcpy(st->hdr, &h, HEADER_SIZE);  // clean even seq=0

            st->sub = create_generic_subscription(
                st->topic, st->type_name, rclcpp::QoS(1),
                [this, st](std::shared_ptr<rclcpp::SerializedMessage> msg) {
                    on_msg(st, msg);
                });

            RCLCPP_INFO(get_logger(), "[%s] %s (%s) -> /dev/shm/%s_*",
                        st->name.c_str(), st->topic.c_str(),
                        st->type_name.c_str(), st->name.c_str());
            streams_.push_back(st);
        }
    }

private:
    void on_msg(std::shared_ptr<Stream> st,
                std::shared_ptr<rclcpp::SerializedMessage> msg) {
        const auto& rcl = msg->get_rcl_serialized_message();
        size_t n = rcl.buffer_length;
        if (n > st->frame_cap) {
            RCLCPP_WARN(get_logger(), "[%s] payload %zu > cap %zu, dropped",
                        st->name.c_str(), n, st->frame_cap);
            return;
        }
        uint64_t ts = std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::system_clock::now().time_since_epoch()).count();

        // ---- SEQLOCK write (writer never blocks; PROMPT.md §5) ----
        st->seq += 1;                               // odd: writing
        store_seq_release(st->hdr, st->seq);

        std::memcpy(st->frame, rcl.buffer, n);

        Header h{};
        h.seq = st->seq;
        h.encoding_id = ENC_CDR;
        h.data_size = static_cast<uint32_t>(n);
        h.timestamp_ns = ts;
        set_type_name(h, st->type_name);
        std::memcpy(st->hdr, &h, HEADER_SIZE);

        write_recipe(st->recipe_path, st->type_name, n, ts);

        st->seq += 1;                               // even: stable
        store_seq_release(st->hdr, st->seq);
    }

    static void write_recipe(const std::string& path, const std::string& type_name,
                             size_t n, uint64_t ts) {
        std::ostringstream o;
        o << "{\"type_name\":\"" << type_name << "\",\"encoding\":\"cdr\","
          << "\"data_size\":" << n << ",\"timestamp_ns\":" << ts << "}";
        std::string tmp = path + ".tmp";
        { std::ofstream f(tmp); f << o.str(); }
        std::rename(tmp.c_str(), path.c_str());     // atomic swap
    }

    size_t max_bytes_ = 0;
    std::vector<std::shared_ptr<Stream>> streams_;
};

int main(int argc, char** argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<GenericSHMWriter>());
    rclcpp::shutdown();
    return 0;
}

File: src/shm_bridge_python/shm_bridge_python/generic_shm_writer.py
#!/usr/bin/env python3
"""
generic_shm_writer.py — type-agnostic ROS 2 -> shared memory writer (PROMPT.md §5).

Subscribes to one or more topics (configured by parameter). For each message it
picks an encoding:
  FLAT  — zero-copy numeric buffer for image-like messages (Image/PointCloud2/...)
  CDR   — rclpy.serialization for everything else (universal, covers all types)
then publishes header + frame + JSON recipe under a seqlock so a type-agnostic
reader can reconstruct the message.

Parameters:
  streams (string array)  list of "stream_name:topic:Type", e.g.
        ["rgb:/rgb:sensor_msgs/msg/Image", "tf:/tf:tf2_msgs/msg/TFMessage"]
  max_bytes (int)         per-stream frame buffer cap (default 1920*1080*4)
"""
import time

import numpy as np
import rclpy
from rclpy.node import Node
from rclpy.serialization import serialize_message
from rosidl_runtime_py.utilities import get_message

from shm_bridge_python import shm_contract as C

# Types we encode as a contiguous numeric buffer (FLAT). Everything else -> CDR.
# Kept deliberately narrow and explicit (PROMPT.md §11 open question).
FLAT_HANDLERS = {}


def _flat_image(msg):
    """sensor_msgs/Image -> (buffer, width, height, channels, dtype)."""
    h, w = msg.height, msg.width
    data = np.frombuffer(bytes(msg.data), dtype=np.uint8)
    ch = (data.size // (h * w)) if (h * w) else 1
    return data, w, h, max(ch, 1), "uint8"


def _flat_pointcloud2(msg):
    """sensor_msgs/PointCloud2 -> raw point buffer (uint8), shape kept in recipe."""
    data = np.frombuffer(bytes(msg.data), dtype=np.uint8)
    w = msg.width
    h = msg.height if msg.height > 0 else 1
    return data, w, h, msg.point_step, "uint8"


FLAT_HANDLERS["sensor_msgs/msg/Image"] = _flat_image
FLAT_HANDLERS["sensor_msgs/msg/PointCloud2"] = _flat_pointcloud2


class Stream:
    def __init__(self, name, topic, type_name, max_bytes):
        self.name = name
        self.topic = topic
        self.type_name = type_name
        self.msg_cls = get_message(type_name)

        hp, fp, rp = C.shm_paths(name)
        _, self.hdr_mm, _ = C.open_mmap(hp, C.HEADER_SIZE, create=True)
        _, self.frame_mm, self.frame_cap = C.open_mmap(fp, max_bytes, create=True)
        self.recipe_path = rp
        self.seq = 0
        # initialise header to a clean even seq
        self.hdr_mm[0:C.HEADER_SIZE] = C.pack_header(
            0, C.ENC_CDR, 0, 0, 0, 0, 0, 0, type_name)
        self.hdr_mm.flush()


class GenericSHMWriter(Node):
    def __init__(self):
        super().__init__("generic_shm_writer")
        self.declare_parameter("streams", ["rgb:/rgb:sensor_msgs/msg/Image"])
        self.declare_parameter("max_bytes", 1920 * 1080 * 4)

        max_bytes = int(self.get_parameter("max_bytes").value)
        specs = self.get_parameter("streams").value
        self.streams = {}

        for spec in specs:
            # split into exactly 3 parts; topic may not contain ':' (ROS names don't)
            name, topic, type_name = spec.split(":", 2)
            st = Stream(name, topic, type_name, max_bytes)
            self.streams[name] = st
            self.create_subscription(
                st.msg_cls, topic,
                lambda msg, s=st: self._on_msg(s, msg), 1)
            self.get_logger().info(
                f"[{name}] {topic} ({type_name}) -> /dev/shm/{name}_*")

    def _on_msg(self, st: Stream, msg):
        # ---- choose encoding ----
        handler = FLAT_HANDLERS.get(st.type_name)
        if handler is not None:
            buf, width, height, channels, dtype = handler(msg)
            payload = buf.tobytes()
            encoding_id = C.ENC_FLAT
            dtype_id = C.ID_BY_DTYPE.get(dtype, 0)
            recipe = {
                "type_name": st.type_name, "encoding": "flat", "dtype": dtype,
                "shape": [height, width, channels] if channels > 1 else [height, width],
                "data_size": len(payload),
            }
        else:
            payload = bytes(serialize_message(msg))
            encoding_id = C.ENC_CDR
            width = height = channels = dtype_id = 0
            recipe = {
                "type_name": st.type_name, "encoding": "cdr",
                "data_size": len(payload),
            }

        if len(payload) > st.frame_cap:
            self.get_logger().warn(
                f"[{st.name}] payload {len(payload)} > cap {st.frame_cap}, dropped")
            return

        ts = time.time_ns()
        recipe["timestamp_ns"] = ts

        # ---- SEQLOCK write (writer never blocks; PROMPT.md §5) ----
        st.seq += 1                                   # -> odd: writing
        st.hdr_mm[0:4] = (st.seq).to_bytes(4, "little")
        st.hdr_mm.flush(0, C.HEADER_SIZE)

        st.frame_mm[0:len(payload)] = payload
        st.hdr_mm[0:C.HEADER_SIZE] = C.pack_header(
            st.seq, encoding_id, len(payload), width, height,
            channels, dtype_id, ts, st.type_name)
        C.write_recipe(st.recipe_path, recipe)

        st.seq += 1                                   # -> even: stable
        st.hdr_mm[0:4] = (st.seq).to_bytes(4, "little")


def main():
    rclpy.init()
    node = GenericSHMWriter()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
